src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -103,21 +104,37 @@ public ImmutableArray<PackageSource> GetPackageSources()
                 return packageSources;
             }
 
-            try
+            if (_packageSourcesAsync is null)
             {
-                packageSources = _packageSourceProvider.Value.GetSources(includeUnOfficial: true, includeDisabled: false)
-                    .SelectAsArray(r => new PackageSource(r.Key, r.Value));
+                lock (_gate)
+                {
+                    if (_packageSourcesAsync is null)
+                    {
+                        _packageSourcesAsync = ThreadingContext.JoinableTaskFactory.RunAsync(() => GetPackageSourcesImplAsync());
+                    }
+                }
             }
-            catch (Exception ex) when (ex is InvalidDataException || ex is InvalidOperationException)
+
+            if (allowSwitchToMainThread)
             {
-                // These exceptions can happen when the nuget.config file is broken.
-                packageSources = ImmutableArray<PackageSource>.Empty;
+                using var combinedToken = _tokenSource.Token.CombineWith(cancellationToken);
+                packageSources = await _packageSourcesAsync.JoinAsync(combinedToken.Token).ConfigureAwait(false);
             }
-            catch (ArgumentException ae) when (FatalError.ReportWithoutCrash(ae))
+            else if (_packageSourcesAsync.IsCompleted)
             {
-                // This exception can happen when the nuget.config file is broken, e.g. invalid credentials.
-                // https://github.com/dotnet/roslyn/issues/40857
-                packageSources = ImmutableArray<PackageSource>.Empty;
+                if (_packageSourcesAsync.Task.Status != TaskStatus.RanToCompletion)
+                {
+                    // The operation did not complete successfully
+                    return null;
+                }
+
+                packageSources = _packageSourcesAsync.GetAwaiter().GetResult();
```

Comment:
do we need the above 'if' check?  normally we would 'await' this, meaning we would get exceptions. so .GetAwaiter().GetResult() in all cases seems reasonable.

Reply:
Makes sense. The check was added prior to the automatic retry in `GetPackageSourcesImplAsync`.

Reply:
Simplified to use the `await` for this scenario.